src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -34,242 +34,128 @@ private sealed class Parser
-                            if (compilationSemanticModel.GetSymbolInfo(attributeSyntax, cancellationToken).Symbol is not IMethodSymbol attributeSymbol)
-                            {
-                                continue;
-                            }
-
-                            INamedTypeSymbol attributeContainingTypeSymbol = attributeSymbol.ContainingType;
-
-                            if (jsonSerializableAttributeSymbol.Equals(attributeContainingTypeSymbol, SymbolEqualityComparer.Default))
-                            {
-                                (serializableAttributeList ??= new List<AttributeSyntax>()).Add(attributeSyntax);
-                            }
-                            else if (jsonSourceGenerationOptionsAttributeSymbol.Equals(attributeContainingTypeSymbol, SymbolEqualityComparer.Default))
-                            {
-                                options = GetSerializerOptions(attributeSyntax);
-                            }
-                        }
-
-                        if (serializableAttributeList == null)
-                        {
-                            // No types were indicated with [JsonSerializable]
-                            continue;
-                        }
+                Debug.Assert(_knownSymbols.JsonSerializerContextType != null);
 
-                        INamedTypeSymbol? contextTypeSymbol = compilationSemanticModel.GetDeclaredSymbol(classDeclarationSyntax, cancellationToken);
-                        Debug.Assert(contextTypeSymbol != null);
+                // Ensure context-scoped metadata caches are empty.
+                Debug.Assert(_typesToGenerate.Count == 0);
+                Debug.Assert(_generatedTypes.Count == 0);
 
-                        Location contextLocation = contextTypeSymbol.Locations.Length > 0 ? contextTypeSymbol.Locations[0] : Location.None;
-
-                        if (!TryGetClassDeclarationList(contextTypeSymbol, out List<string>? classDeclarationList))
-                        {
-                            // Class or one of its containing types is not partial so we can't add to it.
-                            ReportDiagnostic(ContextClassesMustBePartial, contextLocation, new string[] { contextTypeSymbol.Name });
-                            continue;
-                        }
+                if (!DerivesFromJsonSerializerContext(contextClassDeclaration, _knownSymbols.JsonSerializerContextType, semanticModel, cancellationToken))
+                {
+                    return null;
+                }
 
-                        options ??= new JsonSourceGenerationOptionsAttribute();
+                if (!TryParseJsonSerializerContextAttributes(
+                        contextClassDeclaration,
+                        semanticModel,
+                        cancellationToken,
+                        out List<TypeToGenerate>? rootSerializableTypes,
+                        out JsonSourceGenerationOptionsAttribute? options))
```

Comment:
nit: indenting?

Reply:
Or 4 spaces rather than 8?  It's not a big deal, it just doesn't match the rest of the file's indenting.